sdk/go/common/testing/iotest/log_writer.go

```diff
@@ -0,0 +1,88 @@
+package iotest
+
+import (
+	"bytes"
+	"io"
+	"sync"
+	"testing"
+)
+
+// logWriter is an io.Writer that writes to a testing.T.
+type logWriter struct {
+	// t holds the subset of testing.TB
+	// that we are allowed to use.
+	//
+	// We're not storing testing.TB directly to ensure that
+	// we don't accidentally use other log methods.
+	t interface {
+		Logf(string, ...interface{})
+		Helper()
+	}
+
+	// Holds buffered text for the next write or flush
+	// if we haven't yet seen a newline.
+	buff bytes.Buffer
+	mu   sync.Mutex // guards buff
+}
+
+var _ io.Writer = (*logWriter)(nil)
+
+// LogWriter builds and returns an io.Writer that
+// writes messages to the given testing.TB.
+// It ensures that each line is logged separately.
+//
+// Any trailing buffered text that does not end with a newline
+// is flushed when the test finishes.
+//
+// The returned writer is safe for concurrent use
+// from multiple parallel tests.
+func LogWriter(t testing.TB) io.Writer {
+	w := logWriter{t: t}
+	t.Cleanup(w.flush)
+	return &w
+}
+
+func (w *logWriter) Write(bs []byte) (int, error) {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	w.t.Helper() // so that the log message points to the caller
```

Comment:
These are the primary changes to this from the original version:
thread safety and marking it as a helper.